x-pack/agent/pkg/agent/application/fleet_gateway.go

```diff
@@ -0,0 +1,112 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package application
+
+import (
+	"time"
+
+	"github.com/elastic/beats/x-pack/agent/pkg/core/logger"
+	"github.com/elastic/beats/x-pack/agent/pkg/fleetapi"
+	"github.com/elastic/beats/x-pack/agent/pkg/scheduler"
+)
+
+type dispatcher interface {
+	Dispatch(...action) error
+}
+
+type fleetGateway struct {
+	log        *logger.Logger
+	dispatcher dispatcher
+	client     clienter
+	scheduler  scheduler.Scheduler
+	agentID    string
+	done       chan struct{}
+}
+
+type fleetGatewaySettings struct {
+	Duration time.Duration
+}
+
+func newFleetGateway(
+	log *logger.Logger,
+	settings *fleetGatewaySettings,
+	agentID string,
+	client clienter,
+	d dispatcher,
+) (*fleetGateway, error) {
+	scheduler := scheduler.NewPeriodic(settings.Duration)
+	return newFleetGatewayWithScheduler(
+		log,
+		settings,
+		agentID,
+		client,
+		d,
+		scheduler,
+	)
+}
+
+func newFleetGatewayWithScheduler(
+	log *logger.Logger,
+	settings *fleetGatewaySettings,
+	agentID string,
+	client clienter,
+	d dispatcher,
+	scheduler scheduler.Scheduler,
+) (*fleetGateway, error) {
+	return &fleetGateway{
+		log:        log,
+		dispatcher: d,
+		client:     client,
+		agentID:    agentID, //TODO(ph): this need to be a struct.
+		scheduler:  scheduler,
+		done:       make(chan struct{}),
+	}, nil
+}
+
+func (f *fleetGateway) worker() {
+	for {
+		select {
+		case <-f.scheduler.WaitTick():
+			f.log.Debug("FleetGateway calling Checkin API")
+			resp, err := f.execute()
+			if err != nil {
+				f.log.Error(err)
+				continue
+			}
+
+			actions := make([]action, len(resp.Actions))
```

Comment:
why do you need to repack this?

Reply:
Because the `actionDispatcher` doesn't have hard dependencies on `fleetapi.Action`